crates/store/src/state/state_db.rs

```diff
@@ -94,152 +291,342 @@ impl<'a> StateTree<'a> {
-        self.tree.update(key, value)?;
-        // record block's kv state
-        match self.context {
-            StateContext::AttachBlock(block_number) => {
-                self.db()
-                    .record_block_state(block_number, key, value)
-                    .expect("record block state");
-            }
-            StateContext::DetachBlock(_) => {
-                // ignore
-            }
-            ctx => {
-                panic!("wrong ctx: {:?}", ctx);
-            }
-        }
-        if log_enabled!(log::Level::Trace) {
-            let k: Byte32 = key.pack();
-            let v: Byte32 = value.pack();
-            log::trace!(
-                "[state-trace] update_raw ctx:{:?} k:{} v:{}",
-                self.context,
-                k,
-                v
-            );
+        if let Some(tracker) = self.state_tracker.as_ref() {
+            tracker.touch_key(&key);
         }
+        self.journal.push(JournalEntry::UpdateRaw {
+            key,
+            prev_value: self.get_raw(&key)?,
+        });
+        self.dirty_state.insert(key, value);
         Ok(())
     }
 
     fn get_account_count(&self) -> Result<u32, StateError> {
-        if log_enabled!(log::Level::Trace) {
-            log::trace!(
-                "[state-trace] get_account_count ctx:{:?} count:{}",
-                self.context,
-                self.account_count
-            );
+        if let Some(count) = self.dirty_account_count {
+            return Ok(count);
         }
-        Ok(self.account_count)
+        self.state.get_account_count()
     }
 
     fn set_account_count(&mut self, count: u32) -> Result<(), StateError> {
-        if log_enabled!(log::Level::Trace) {
-            log::trace!(
-                "[state-trace] set_account_count ctx:{:?} origin: {} count:{}",
-                self.context,
-                self.account_count,
-                count
-            );
-        }
-        self.account_count = count;
+        self.journal.push(JournalEntry::SetAccountCount {
+            prev_count: self.get_account_count()?,
+        });
+        self.dirty_account_count = Some(count);
         Ok(())
     }
 
+    /// calculate root
     fn calculate_root(&self) -> Result<H256, StateError> {
-        let root = self.tree.root();
-        Ok(*root)
+        // finalise dirty state
+        if self.is_dirty() {
+            return Err(StateError::Store);
+        }
+        self.state.calculate_root()
     }
 }
 
-impl<'a> CodeStore for StateTree<'a> {
+impl<S: CodeStore> CodeStore for StateDB<S> {
     fn insert_script(&mut self, script_hash: H256, script: packed::Script) {
-        self.db()
-            .insert_raw(COLUMN_SCRIPT, script_hash.as_slice(), script.as_slice())
-            .expect("insert script");
+        self.journal.push(JournalEntry::InsertScript {
+            script_hash,
+            prev_exist: self.get_script(&script_hash).is_some(),
```

Comment:
Would `self.dirty_scripts.contains(&script_hash)` be better?

Otherwise if the script exists in self.state, but not in self.dirty_scripts, after revert it would still be in dirty scripts.